apps/aecore/src/aec_pow_cuckoo.erl

```diff
@@ -184,6 +191,47 @@ verify(Hash, Nonce, Soln) ->
             {error, {unknown, {C, E}}}
     end.
 
+%%------------------------------------------------------------------------------
+%% @doc
+%%   Creates the 80-byte Cuckoo buffer (hex encoded) from a base64-encoded
+%%   hash and a uint64 nonce.
+%% @end
+%%------------------------------------------------------------------------------
+-spec pack_header_and_nonce(binary(), aec_pow:nonce()) -> string().
+pack_header_and_nonce(Hash, Nonce) ->
+    %% Cuckoo originally uses 32-bit nonces inserted at the end of its 80-byte buffer.
+    %% This buffer is hashed into the keys used by the main algorithm.
+    %%
+    %% We intend to insert our 64-bit Nonce before the location Cuckoo writes its
+    %% 32-bit nonce into. We cannot do it directly because Cuckoo supports that
+    %% (with option -x) for the mean miner only. Therefore, we base64-encode both the
+    %% hash of the block header and the nonce and pass the resulting command-line
+    %% friendly string with the -h option to Cuckoo.
+    %%
+    %% The SHA256 hash is 32 bytes (44 chars base64-encoded), the nonce is 8 bytes
+    %% (12 chars base64-encoded). There is still room for zero characters between hash
+    %% and the nonce (cannot use zero bytes that Cuckoo normally uses for padding).
+    %%
+    %% (Base64 encoding: see RFC 3548, Section 3:
+    %% https://tools.ietf.org/html/rfc3548#page-4
+    %% converts every triplet of bytes to 4 characters: from N bytes to 4*ceil(N/3)
+    %% bytes.)
+    %%
+    %% Like Cuckoo, (and unlike we hash the header) we use little-endian here.
+    NonceStr = base64:encode_to_string(<<Nonce:64/little-unsigned-integer>>),
+    HashStr = base64:encode_to_string(Hash),
+    12 = length(NonceStr),
+    44 = length(HashStr),
+    %% Fill in space between the hash and the nonce so that is spans the
+    %% 76-bytes before the 4 bytes Cuckoo overwrites. 76 - 44 - 12 = 20.
```

Comment:
This assumes that Cuckoo miner binary overwrites last 4 bytes of the 80 bytes. Links to assumption (mean and lean) would help reader here. What if actually it overwrites 8 bytes hence eating up some of the nonce?

Also - why attempting to use the end of the buffer? I.e. why not avoiding this 20 bytes altogether?

Reply:
```c
((u32 *)headernonce)[len/sizeof(u32)-1] = htole32(nonce); // place nonce at end
```

Again this is Akos design... I would probably have put it immediately after the Hash but I didn't have a strong enough opinion to make a change. It is quickly fixed if you think it is much better?

Reply:
I had a word with Akos - it was for placing the nonce as close as possible to where the upstream Cuckoo miner binary places it. So this 20-bytes interval can be dropped, I suggest to do so for this PR (so it makes describing the PoW in the wiki easier).

Reply:
Good, I'll move it to the `obvious` place then! I will also make a comment that lean/mean (and verify) will put a nonce = 0 at the end of its buffer...